### PR TITLE
Added Search for Hex functionality to the unified SearchReplace dialog.

### PR DESCRIPTION
**Pull request by MKadaner**
## Summary
Added Search for Hex functionality to the unified SearchReplace dialog.

 
## References
- #605

## Checklist
- [X] I have followed the [contributing guidelines](https://github.com/FarGroup/FarManager/blob/master/CONTRIBUTING.md).
- [X] I have discussed this with project maintainers: #584 <br/>
If not checked, I accept that this work might be rejected in favor of a different great big ineffable plan.<br/>

### CODE REVIEW

**Comment by MKadaner** [comment]
@rohitab, Could you please test this draft PR? There are five things to test:
- The behavior of Editor Search / Replace dialog. For testing purposes, I enabled Hex edit controls. It's UI/UX only; there is no attempt to actually implement Hex search in Editor.
- The behavior of Help Search dialog.
- The search functionality in Help. The changes in this area were trivial, so it does not require much attention.

Comment out Hex controls in Editor Search / Replace dialog here: https://github.com/MKadaner/FarManager/blob/8680b5639d6b3359674a156d0fbc86a84429e151/far/editor.cpp#L156

Rebuild and test:
- The behavior of Editor Search / Replace dialog without Hex controls; this is how it should be in production.
- The search functionality in Editor. I did some refactoring here; could introduce bugs.

Thank you in advance!

**Comment by MKadaner** [comment]
@alabuzhev, Could you please look at this change? It is practically ready except for the testing.

File: far/findfile.cpp
Comment (alabuzhev) [inline far/findfile.cpp:3237]: Removing this looks logical, but what was it for?

File: far/stddlg.cpp
Comment (alabuzhev) [inline far/stddlg.cpp:179]: All these can be const supposedly?

File: far/stddlg.cpp
Comment (alabuzhev) [inline far/stddlg.cpp:260]: Minor: please format as in other places, e.g.
```C++
switch (...)
{
case a:
    ...
    break;

case b:
    {
        ...
    }
    break;

    ...
}
```

File: far/stddlg.cpp
Comment (alabuzhev) [inline far/stddlg.cpp:362]: - Minor: please add empty lines between `if`s, it's somewhat hard to read one big block.

- Something like this can probably be more readable too:
```C++
const auto set_if = [&](auto& Param, int const Id)
{
	if (Param.has_value())
		Param = DlgItems[Id].Selected == BSTATE_CHECKED;
};

set_if(Params.Hex, dlg_radio_hex);
set_if(Params.CaseSensitive, dlg_checkbox_case);
...
```

File: far/stddlg.hpp
Comment (alabuzhev) [inline far/stddlg.hpp:61]: This is used only in Editor, in all other places it has to be hidden explicitly, maybe make it true by default, or rather invert as `ShowButtonAll = false`?

File: far/editor.cpp
Comment (MKadaner) [reply far/editor.cpp:129]: It will introduce new dependency, global on strmix (something) and makes global aware of codepage (though in a trivial way). I was not sure it would be a good idea. Otherwise, these functions should be in global.

File: far/editor.cpp
Comment (MKadaner) [reply far/editor.cpp:156]: That's the way to debug / test new Hex functionality in the Search / Replace dialog, mostly to get meaningful feedback from @rohitab. This is why the PR is a draft. I will remove this line when he is done. The Editor Search should not change for now. Adding Hex search to Editor is a completely different story.

File: far/stddlg.hpp
Comment (MKadaner) [reply far/stddlg.hpp:61]: Invert "Button All" flag only, right? "Replace Mode" is fine? I thought about it, all name variants I thought of are ambiguous in the sense that it is unclear whether we are talking about a particular button, or about all of something. After spending rather much time on these considerations, I decided to keep original meaning at least. I will happily flip it to make default init more natural.

**Comment by MKadaner** [comment]
> It is rather massive so probably will take a few days.

Sure!

It is as small as I could make it. Originally, I wanted to use the dialog in Viewer as well, but then realized the change was already too big.

File: far/findfile.cpp
Comment (MKadaner) [reply far/findfile.cpp:3237]: _Global static_ `ReplaceMode` in Editor used to control the behavior of `Shif+F7` / `Alt+F7`: either next find or next replace, _for all Editors!_ So, in the latest versions, after finding some files, all Editors would do "next find." I _guess_ this was needed in the scenario when the user opens a file for edit from the list of found files and presses `Shif+F7` expecting to _find_ the "containing" pattern. If not these lines, the pattern would be replaced with whatever happened to be the last replace string. I vaguely remember that some time ago the list of found files would switch to the existing Editor instance, if the user clicks `F4` on the file which was already being edited. Or at least it would ask the user and _allow_ to switch to the existing instance. Recently, the list of found files only allows to launch the new Editor instance. Once again, this is my educated guess, I may be wrong. If somebody knows, please chime in.

Assuming the above is correct, I removed the global `ReplaceMode` and use the flag stored in the `SearchReplaceDlgParams` instance owned by the Editor. If this breaks somebody's expectations, or I simply screwed something up, I will fix the behavior according to the actual feature request instead of trying to guess what might be useful.

File: far/editor.cpp
Comment (MKadaner) [reply far/editor.cpp:82]: 🙂 I made this comment (and the other one), instead of removing the whole thing, as a reminder to discuss the behavioral consequences and desirability of replacing global statics with instance variables. It was long before creating the PR, so there was no other place to make this note. Then I kinda forgot about it, the PR being a draft.

So, looks like we are OK with this change. I will remove the dead code for the next iteration.

File: far/editor.cpp
Comment (MKadaner) [reply far/editor.cpp]: Moved. If you think the dependency should not be there, let me know, I'll move them back to Editor.

File: far/editor.cpp
Comment (MKadaner) [reply far/editor.cpp]: Added spaces here and on surrounding lines. Seems to be more prevalent style in this function (file?)

File: far/editor.cpp
Comment (MKadaner) [reply far/editor.cpp]: I do like the curly syntax—purely personal, of course. To me, equal sign suggests assignment. Having a dedicated syntax for definition was a great language improvement.

Changed it here but kept curlies in many other places. If you do not insist, that is.

File: far/stddlg.cpp
Comment (MKadaner) [reply far/stddlg.cpp]: Yeah! Thrown in `const` liberally.

File: far/stddlg.cpp
Comment (MKadaner) [reply far/stddlg.cpp]: Sure, Done. In fact, I was somewhat confused since I noticed different style in different places. Having `break` on the same line as closing curly brace is my invention though. :)

File: far/editor.cpp
Comment (alabuzhev) [reply far/editor.cpp]: Should be ok. Even if we later discover that we want the shared state back, it would be better to make it more obvious than it was.

File: far/editor.cpp
Comment (alabuzhev) [reply far/editor.cpp]: Thanks. I mean `[ = ]` is preferable, `[=]` is fine too, but `[= ]` looks weird :)

File: far/editor.cpp
Comment (alabuzhev) [reply far/editor.cpp]: Thanks

File: far/findfile.cpp
Comment (alabuzhev) [reply far/findfile.cpp:3237]: Thanks. 

> Or at least it would ask the user and allow to switch to the existing instance. Recently, the list of found files only allows to launch the new Editor instance.

It does allow that when invoked from panels, but not from find file unfortunately, mainly because such editor instance has to be modal. Apparently there were technical issues with modalizing existing windows in the past, I don't know how relevant this limitation is these days.

File: far/stddlg.hpp
Comment (alabuzhev) [reply far/stddlg.hpp]: Thanks

File: far/findfile.cpp
Comment (MKadaner) [reply far/findfile.cpp:3237]: When an Editor is opened from panels, the confusing scenario with unexpected replacement is moot, so we should be fine here.

**Comment by rohitab** [comment]
@MKadaner I just built and tested version `3.0.6093.0 (Private) x64` from [MKadaner:mzk-stddlg-search](https://github.com/MKadaner/FarManager/tree/mzk-stddlg-search)

I tested out the changes you made and they seem to be working fine. The only issue I found so far is related to a hotkey in the _Replace_ dialog of the Editor. If you press `Alt+R`, instead of clicking the _Replace_ button, it inserts the word under the cursor in the _Search for_ box. Basically, `Alt+R` works the same as `Alt+O`. In your test version, the one with Hex controls, pressing `Alt+R` selects the _Text_/_Hex_ radio button.

You mentioned that Hex search is UI only and not implemented in Editor, so I'm not going to report any issues related to those. As an example, in the _Replace_ dialog in the Editor, if you select _Hex_ search, the _Replace with_ box accepts text input, not hex input.

The focus issue https://github.com/FarGroup/FarManager/issues/605#issuecomment-1376809669 is still present in this release. However, it's not present in the new Hex control that you added in the Editor for testing. The cursor position issue mentioned in #607 is also present in both the Viewer and the new Hex control in the Editor. I'm not sure if you included a fix for that, so I'm reporting it just in case.

I did notice a few minor issues with Help and Hex search in Viewer, but they are present in the main _Far_ branch, and are not related to your changes. I will open separate tickets for them to avoid any confusion.

Apologies for not being able to test sooner.

**Comment by MKadaner** [comment]
@rohitab, Thank you so much for testing; incredibly helpful. Never mind the time, nobody's get paid here.

> ... issue ... related to a hotkey in the Replace dialog ...

This is a bug. I will have fixed it in the next iteration (will push in a few minutes).

> The focus issue https://github.com/FarGroup/FarManager/issues/605#issuecomment-1376809669 ...

If it is in the Viewer, that is expected, I did not touch it. Frankly, I do not want to look there. I'd rather concentrate on replacing Viewer Search dialog with the unified one (next after this PR is merged), and the question will be moot.

> The cursor position issue mentioned in #607 is also present in both the Viewer and the new Hex control in the Editor.

Yeah! This one is an issue on its own and is pretty much independent of the dialog behavior. Since it is not a regression, let's keep track of it in #607 and I will do something about it, probably after reworking Viewer Search.

> ... the Replace dialog in the Editor, if you select Hex search, the Replace with box accepts text input, not hex input.

For now, I just did not touch the Replace control, since in production it will never appear together with the Hex controls, not util we implement Hex search in Editor. And this one is interesting. When I try to think through my scenarios, I am not sure whether I want to replace a hex pattern with hex or text string. I need to actually use hex search in Editor to decide. Maybe the replace option should even be independent from the search one? Another interesting question is whether hex search in Editor should scan the decoded text or the underlying file, as in the Viewer. If it searches the file content, should it still search line by line, skipping line ends, as Editor does now?

I think we should start collecting requirements. @alabuzhev et al, does anybody have any suggestions? At any rate, hex search in Editor is relatively low on my priority list.

**Comment by alabuzhev** [comment]
> I think we should start collecting requirements. @alabuzhev et al, does anybody have any suggestions? At any rate, hex search in Editor is relatively low on my priority list.

<details>
<summary>Yes</summary>

image
</details>

I doubt that Hex search in Editor would make much sense - Editor does not work with bytes, so such mode would be quite hard to implement properly, and implementing it halfway would only confuse people.

**Comment by MKadaner** [comment]
> Yes

:) I've actually considered the "should". I remember using Far Editor several times to patch binary files like executables or images, even recently. It's possible if one is careful. I can imagine hex search / replace could be useful in such endeavors. But I agree that this feature is of very low priority, even if it's deemed useful.

**Review by alabuzhev — APPROVED** [review]

**Comment by MKadaner** [comment]
Removed test scaffolding, changelog, version, rebased, squashed, ready.

Thank you for the review!

**Comment by alabuzhev** [comment]
@MKadaner I've pushed a commit before noticing your latest changes, please rebase again.

**Comment by MKadaner** [comment]
@alabuzhev Done!

**Comment by alabuzhev** [comment]
Thank you

**Comment by rohitab** [comment]
**Editor Replace Button Hotkey**

Most dialogs in _Far_ do not have hotkeys for the default button, since you can just press `Enter`. For example, in the _Search_ dialog in Viewer/Editor/Help, the _Search_ button does not have a hotkey. However, in the _Replace_ dialog in the Editor, the _Replace_ button has an `Alt+R` hotkey.

What do you think about removing the hotkey from the _Replace_ button and assigning `Alt+R` to the _Replace with_ edit control instead? It seems natural to me, to use `Alt+S` and `Alt+R` to switch between _Search_ and _Replace_ with boxes. It might inconvenience people that are accustomed to using the `Alt+R`/`Alt+E` hotkeys though.

**Comment by MKadaner** [comment]
I also noticed this discrepancy. I think this is a good idea. The decision is mostly for the language maintainers, except for there is no dedicated maintainer for English localization.

P.S. maybe move this discussion to, well, "Discussions?"

**Comment by rohitab** [comment]
Moved to Discussions: https://github.com/FarGroup/FarManager/discussions/616

**Pull request merged**
